MediaGallerySynchronization/Model/Synchronize.php

```diff
@@ -51,125 +30,72 @@ class Synchronize implements SynchronizeInterface
     private $synchronizerPool;
 
     /**
-     * @var GetAssetsIterator
+     * @var AssetsBatchGenerator
      */
-    private $getAssetsIterator;
+    private $batchGenerator;
+
+    /**
+     * @var int
+     */
+    private $batchSize;
 
     /**
      * @var ResolveNonExistedAssets
      */
     private $resolveNonExistedAssets;
-
+    
     /**
-     * @param IsPathBlacklistedInterface $isPathBlacklisted
-     * @param Filesystem $filesystem
+     * @param ResolveNonExistedAssets $resolveNonExistedAssets
      * @param LoggerInterface $log
      * @param SynchronizerPool $synchronizerPool
-     * @param GetAssetsIterator $assetsIterator
-     * @param ResolveNonExistedAssets $resolveNonExistedAssets
+     * @param AssetsBatchGenerator $batchGenerator
+     * @param int $batchSize
      */
     public function __construct(
-        IsPathBlacklistedInterface $isPathBlacklisted,
-        Filesystem $filesystem,
+        ResolveNonExistedAssets $resolveNonExistedAssets,
         LoggerInterface $log,
         SynchronizerPool $synchronizerPool,
-        GetAssetsIterator $assetsIterator,
-        ResolveNonExistedAssets $resolveNonExistedAssets
+        AssetsBatchGenerator $batchGenerator,
+        int $batchSize
     ) {
-        $this->isPathBlacklisted = $isPathBlacklisted;
-        $this->filesystem = $filesystem;
+        $this->resolveNonExistedAssets = $resolveNonExistedAssets;
         $this->log = $log;
         $this->synchronizerPool = $synchronizerPool;
-        $this->getAssetsIterator = $assetsIterator;
-        $this->resolveNonExistedAssets = $resolveNonExistedAssets;
+        $this->batchGenerator = $batchGenerator;
+        $this->batchSize = $batchSize;
     }
 
     /**
      * @inheritdoc
      */
     public function execute(): void
     {
-        $failedItems = [];
-        $assetsPaths = [];
-
-        /** @var \SplFileInfo $item */
-        foreach ($this->getAssetsIterator->execute($this->getMediaDirectory()->getAbsolutePath()) as $item) {
-            $path = $item->getPath() . '/' . $item->getFilename();
-            if (!$this->isApplicable($path)) {
-                continue;
-            }
-
-            $assetsPaths[] = $this->getRelativePath($path, $item->getFilename());
+        $failed = [];
 
-            foreach ($this->synchronizerPool->get() as $synchronizer) {
-                if ($synchronizer instanceof SynchronizeFilesInterface) {
+        foreach ($this->synchronizerPool->get() as $name => $synchronizer) {
+            if ($synchronizer instanceof SynchronizeFilesInterface) {
+                foreach ($this->batchGenerator->getItems($this->batchSize) as $batch) {
                     try {
-                        $synchronizer->execute([$item]);
+                        $synchronizer->execute($batch);
+                        $this->resolveNonExistedAssets->execute($batch);
```

Comment:
I don't think that `ResolveNonExistedAssets` is able to work with partial batches currently, I think we should extract processing not existing assets to a separate process that should load assets from database in batches and check if they still exist in the filesystem